File: ts/currency.ts
// MCP Currency Exchange Server
// Exchange rates using free APIs (exchangerate.host, frankfurter.app)

import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import { z } from 'zod';

const API_BASE = 'https://api.frankfurter.app';

interface CurrencyResult<T = unknown> {
    success: boolean;
    data?: T;
    error?: string;
}

async function fetchJson<T>(url: string): Promise<CurrencyResult<T>> {
    try {
        const response = await fetch(url, {
            headers: {
                Accept: 'application/json',
                'User-Agent': 'Flopsy-Currency-MCP/1.0',
            },
        });

        if (!response.ok) {
            return { success: false, error: `HTTP ${response.status}: ${response.statusText}` };
        }

        const data = (await response.json()) as T;
        return { success: true, data };
    } catch (err) {
        return { success: false, error: (err as Error).message };
    }
}

function formatResult<T>(result: CurrencyResult<T>): {
    content: Array<{ type: 'text'; text: string }>;
} {
    if (!result.success) {
        return {
            content: [{ type: 'text', text: JSON.stringify({ error: result.error }) }],
        };
    }
    return {
        content: [{ type: 'text', text: JSON.stringify(result.data, null, 2) }],
    };
}

interface LatestRates {
    base: string;
    date: string;
    rates: Record<string, number>;
}

async function getLatestRates(base: string, symbols?: string[]): Promise<CurrencyResult> {
    let url = `${API_BASE}/latest?from=${base.toUpperCase()}`;
    if (symbols && symbols.length > 0) {
        url += `&to=${symbols.map((s) => s.toUpperCase()).join(',')}`;
    }

    const result = await fetchJson<LatestRates>(url);

    if (!result.success) return result;

    const rates = Object.entries(result.data!.rates).map(([currency, rate]) => ({
        currency,
        rate,
        inverse: (1 / rate).toFixed(6),
    }));

    return {
        success: true,
        data: {
            base: result.data!.base,
            date: result.data!.date,
            count: rates.length,
            rates,
            timestamp: new Date().toISOString(),
        },
    };
}

async function convert(from: string, to: string, amount: number): Promise<CurrencyResult> {
    const url = `${API_BASE}/latest?amount=${amount}&from=${from.toUpperCase()}&to=${to.toUpperCase()}`;

    const result = await fetchJson<LatestRates>(url);

    if (!result.success) return result;

    const converted = result.data!.rates[to.toUpperCase()];
    if (converted === undefined) {
        return { success: false, error: `Currency ${to.toUpperCase()} not found` };
    }
    const rate = converted / amount;

    return {
        success: true,
        data: {
            from: from.toUpperCase(),
            to: to.toUpperCase(),
            amount,
            converted,
            rate,
            date: result.data!.date,
            formatted: `${amount} ${from.toUpperCase()} = ${converted.toFixed(2)} ${to.toUpperCase()}`,
        },
    };
}

interface HistoricalRates {
    base: string;
    start_date: string;
    end_date: string;
    rates: Record<string, Record<string, number>>;
}

async function getHistorical(
    base: string,
    target: string,
    startDate: string,
    endDate: string,
): Promise<CurrencyResult> {
    const url = `${API_BASE}/${startDate}..${endDate}?from=${base.toUpperCase()}&to=${target.toUpperCase()}`;

    const result = await fetchJson<HistoricalRates>(url);

    if (!result.success) return result;

    const history = Object.entries(result.data!.rates).map(([date, rates]) => ({
        date,
        rate: rates[target.toUpperCase()],
    }));

    // Calculate stats - filter out undefined values
    const rateValues = history.map((h) => h.rate).filter((r): r is number => r !== undefined);

    if (rateValues.length === 0) {
        return { success: false, error: 'No rate data found for the specified period' };
    }

    const firstRate = rateValues[0]!;
    const lastRate = rateValues[rateValues.length - 1]!;
    const change = ((lastRate - firstRate) / firstRate) * 100;

    return {
        success: true,
        data: {
            base: base.toUpperCase(),
            target: target.toUpperCase(),
            period: { start: startDate, end: endDate },
            dataPoints: history.length,
            summary: {
                startRate: firstRate,
                endRate: lastRate,
                change: change.toFixed(2) + '%',
                high: Math.max(...rateValues).toFixed(4),
                low: Math.min(...rateValues).toFixed(4),
                average: (rateValues.reduce((a, b) => a + b, 0) / rateValues.length).toFixed(4),
            },
            history,
        },
    };
}

interface CurrencyList {
    [key: string]: string;
}

async function getCurrencies(): Promise<CurrencyResult> {
    const result = await fetchJson<CurrencyList>(`${API_BASE}/currencies`);

    if (!result.success) return result;

    const currencies = Object.entries(result.data!).map(([code, name]) => ({
        code,
        name,
    }));

    // Group by common categories
    const major = ['USD', 'EUR', 'GBP', 'JPY', 'CHF', 'AUD', 'CAD', 'CNY'];
    const majorCurrencies = currencies.filter((c) => major.includes(c.code));
    const otherCurrencies = currencies.filter((c) => !major.includes(c.code));

    return {
        success: true,
        data: {
            total: currencies.length,
            major: majorCurrencies,
            other: otherCurrencies,
        },
    };
}

export function createCurrencyMcpServer() {
    const server = new McpServer({
        name: 'currency-server',
        version: '1.0.0',
    });

    server.registerTool(
        'currency_rates',
        {
            title: 'Exchange Rates',
            description: 'Get current exchange rates for a base currency.',
            inputSchema: {
                base: z.string().default('USD').describe("Base currency code (e.g., 'USD', 'EUR')"),
                symbols: z
                    .array(z.string())
                    .optional()
                    .describe("Target currencies (e.g., ['EUR', 'GBP']). Leave empty for all."),
            },
        },
        async ({ base, symbols }) => {
            const result = await getLatestRates(base, symbols);
            return formatResult(result);
        },
    );

    server.registerTool(
        'currency_convert',
        {
            title: 'Convert Currency',
            description: 'Convert an amount from one currency to another.',
            inputSchema: {
                from: z.string().describe("Source currency (e.g., 'USD')"),
                to: z.string().describe("Target currency (e.g., 'EUR')"),
                amount: z.number().describe('Amount to convert'),
            },
        },
        async ({ from, to, amount }) => {
            const result = await convert(from, to, amount);
            return formatResult(result);
        },
    );

    server.registerTool(
        'currency_history',
        {
            title: 'Historical Rates',
            description: 'Get historical exchange rates between two currencies.',
            inputSchema: {
                base: z.string().describe("Base currency (e.g., 'USD')"),
                target: z.string().describe("Target currency (e.g., 'EUR')"),
                startDate: z.string().describe("Start date (YYYY-MM-DD, e.g., '2024-01-01')"),
                endDate: z.string().describe("End date (YYYY-MM-DD, e.g., '2024-12-31')"),
            },
        },
        async ({ base, target, startDate, endDate }) => {
            const result = await getHistorical(base, target, startDate, endDate);
            return formatResult(result);
        },
    );

    server.registerTool(
        'currency_list',
        {
            title: 'List Currencies',
            description: 'Get list of all supported currencies with their names.',
            inputSchema: {},
        },
        async () => {
            const result = await getCurrencies();
            return formatResult(result);
        },
    );

    server.registerTool(
        'currency_multi',
        {
            title: 'Multi Currency Convert',
            description: 'Convert an amount to multiple currencies at once.',
            inputSchema: {
                from: z.string().describe("Source currency (e.g., 'USD')"),
                amount: z.number().describe('Amount to convert'),
                to: z.array(z.string()).describe("Target currencies (e.g., ['EUR', 'GBP', 'JPY'])"),
            },
        },
        async ({ from, amount, to }) => {
            const url = `${API_BASE}/latest?amount=${amount}&from=${from.toUpperCase()}&to=${to.map((s) => s.toUpperCase()).join(',')}`;
            const result = await fetchJson<LatestRates>(url);

            if (!result.success) return formatResult(result);

            const conversions = Object.entries(result.data!.rates).map(([currency, value]) => ({
                currency,
                value: value.toFixed(2),
                formatted: `${value.toFixed(2)} ${currency}`,
            }));

            return formatResult({
                success: true,
                data: {
                    from: from.toUpperCase(),
                    amount,
                    date: result.data!.date,
                    conversions,
                },
            });
        },
    );

    return server;
}

async function main() {
    const server = createCurrencyMcpServer();

    const transport = new StdioServerTransport();
    await server.connect(transport);
    console.error('[MCP] Currency started');
}

// Run if executed directly
const isMain = import.meta.url === `file://${process.argv[1]}`;
if (isMain) {
    main().catch((error) => {
        console.error('[MCP Currency Server] Fatal error:', error);
        process.exit(1);
    });
}

File: agent/tools/currency.py
"""Currency exchange tool using the free Frankfurter API.

Provides exchange rates, conversion, historical rates, and currency listing.
No API key required.
"""

import asyncio
import json
from concurrent.futures import ThreadPoolExecutor
from typing import ClassVar, Type

import httpx
from cachetools import TTLCache
from langchain_core.tools import BaseTool
from pydantic import BaseModel, Field, PrivateAttr

from agent.logger import get_logger

logger = get_logger("agent.tools.currency", component="currency")

API_BASE = "https://api.frankfurter.app"
HEADERS = {"Accept": "application/json", "User-Agent": "OpenSentinel-Currency/1.0"}


# =============================================================================
# Input Schema
# =============================================================================


class CurrencyInput(BaseModel):
    action: str = Field(
        ...,
        description=(
            "Action to perform: "
            "'rates' — get current exchange rates for a base currency, "
            "'convert' — convert an amount from one currency to another, "
            "'history' — get historical rates between two currencies, "
            "'list' — list all supported currencies, "
            "'multi' — convert an amount to multiple currencies at once."
        ),
    )
    base: str = Field(default="USD", description="Base currency code (e.g., 'USD', 'EUR').")
    target: str = Field(default="", description="Target currency code for 'convert' or 'history'.")
    targets: str = Field(
        default="",
        description="Comma-separated target currencies for 'rates' or 'multi' (e.g., 'EUR,GBP,JPY').",
    )
    amount: float = Field(default=1.0, description="Amount to convert (default: 1.0).")
    start_date: str = Field(default="", description="Start date for 'history' (YYYY-MM-DD).")
    end_date: str = Field(default="", description="End date for 'history' (YYYY-MM-DD).")


# =============================================================================
# Tool Implementation
# =============================================================================


class CurrencyTool(BaseTool):
    name: str = "currency"
    description: str = (
        "Get currency exchange rates, convert amounts, and view historical rates. "
        "Uses the free Frankfurter API (no API key required). "
        "Supports all major currencies.\n\n"
        "Examples:\n"
        "- Current rates: action='rates', base='USD', targets='EUR,GBP,JPY'\n"
        "- Convert: action='convert', base='USD', target='EUR', amount=100\n"
        "- Historical: action='history', base='USD', target='EUR', start_date='2024-01-01', end_date='2024-12-31'\n"
        "- List currencies: action='list'\n"
        "- Multi-convert: action='multi', base='USD', targets='EUR,GBP,JPY', amount=100"
    )
    args_schema: Type[BaseModel] = CurrencyInput
    handle_tool_error: bool = True

    _cache: TTLCache = PrivateAttr()
    _client: httpx.AsyncClient = PrivateAttr()

    MAX_CONTENT_CHARS: ClassVar[int] = 3000

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self._cache = TTLCache(maxsize=200, ttl=600)  # 10 min cache
        self._client = httpx.AsyncClient(headers=HEADERS, timeout=15.0)

    async def _fetch(self, url: str) -> dict:
        """Fetch JSON from Frankfurter API."""
        cache_key = url
        if cache_key in self._cache:
            return self._cache[cache_key]

        response = await self._client.get(url)
        response.raise_for_status()
        data = response.json()
        self._cache[cache_key] = data
        return data

    async def _get_rates(self, base: str, targets: str) -> str:
        url = f"{API_BASE}/latest?from={base.upper()}"
        if targets:
            symbols = ",".join(s.strip().upper() for s in targets.split(",") if s.strip())
            url += f"&to={symbols}"

        data = await self._fetch(url)
        rates = [
            {"currency": code, "rate": rate, "inverse": f"{1/rate:.6f}"}
            for code, rate in data.get("rates", {}).items()
        ]

        return json.dumps({
            "base": data.get("base"),
            "date": data.get("date"),
            "count": len(rates),
            "rates": rates,
        }, indent=2)

    async def _convert(self, base: str, target: str, amount: float) -> str:
        if not target:
            return json.dumps({"error": "Provide a target currency (e.g., target='EUR')"})

        url = f"{API_BASE}/latest?amount={amount}&from={base.upper()}&to={target.upper()}"
        data = await self._fetch(url)

        converted = data.get("rates", {}).get(target.upper())
        if converted is None:
            return json.dumps({"error": f"Currency '{target.upper()}' not found"})

        rate = converted / amount if amount else 0

        return json.dumps({
            "from": base.upper(),
            "to": target.upper(),
            "amount": amount,
            "converted": round(converted, 2),
            "rate": round(rate, 6),
            "date": data.get("date"),
            "formatted": f"{amount} {base.upper()} = {converted:.2f} {target.upper()}",
        }, indent=2)

    async def _get_history(self, base: str, target: str, start_date: str, end_date: str) -> str:
        if not target or not start_date or not end_date:
            return json.dumps({
                "error": "Provide target, start_date (YYYY-MM-DD), and end_date (YYYY-MM-DD)"
            })

        url = f"{API_BASE}/{start_date}..{end_date}?from={base.upper()}&to={target.upper()}"
        data = await self._fetch(url)

        history = []
        for date, rates in sorted(data.get("rates", {}).items()):
            rate_val = rates.get(target.upper())
            if rate_val is not None:
                history.append({"date": date, "rate": rate_val})

        rate_values = [h["rate"] for h in history]
        if rate_values:
            first, last = rate_values[0], rate_values[-1]
            change = ((last - first) / first) * 100 if first else 0
            summary = {
                "start_rate": first,
                "end_rate": last,
                "change": f"{change:.2f}%",
                "high": round(max(rate_values), 4),
                "low": round(min(rate_values), 4),
                "average": round(sum(rate_values) / len(rate_values), 4),
            }
        else:
            summary = {}

        return json.dumps({
            "base": base.upper(),
            "target": target.upper(),
            "period": {"start": start_date, "end": end_date},
            "data_points": len(history),
            "summary": summary,
            "history": history,
        }, indent=2)

    async def _list_currencies(self) -> str:
        data = await self._fetch(f"{API_BASE}/currencies")

        currencies = [{"code": code, "name": name} for code, name in data.items()]
        major_codes = {"USD", "EUR", "GBP", "JPY", "CHF", "AUD", "CAD", "CNY"}
        major = [c for c in currencies if c["code"] in major_codes]
        other = [c for c in currencies if c["code"] not in major_codes]

        return json.dumps({
            "total": len(currencies),
            "major": major,
            "other": other,
        }, indent=2)

    async def _multi_convert(self, base: str, targets: str, amount: float) -> str:
        if not targets:
            return json.dumps({"error": "Provide target currencies (e.g., targets='EUR,GBP,JPY')"})

        symbols = ",".join(s.strip().upper() for s in targets.split(",") if s.strip())
        url = f"{API_BASE}/latest?amount={amount}&from={base.upper()}&to={symbols}"
        data = await self._fetch(url)

        conversions = [
            {"currency": code, "value": f"{value:.2f}", "formatted": f"{value:.2f} {code}"}
            for code, value in data.get("rates", {}).items()
        ]

        return json.dumps({
            "from": base.upper(),
            "amount": amount,
            "date": data.get("date"),
            "conversions": conversions,
        }, indent=2)

    # -------------------------------------------------------------------------
    # BaseTool interface
    # -------------------------------------------------------------------------

    async def _arun(
        self,
        action: str,
        base: str = "USD",
        target: str = "",
        targets: str = "",
        amount: float = 1.0,
        start_date: str = "",
        end_date: str = "",
    ) -> str:
        logger.info("currency_action", action=action, base=base, target=target)
        try:
            match action.lower():
                case "rates":
                    return await self._get_rates(base, targets)
                case "convert":
                    return await self._convert(base, target, amount)
                case "history":
                    return await self._get_history(base, target, start_date, end_date)
                case "list":
                    return await self._list_currencies()
                case "multi":
                    return await self._multi_convert(base, targets, amount)
                case _:
                    return json.dumps({
                        "error": f"Unknown action '{action}'. Use: rates, convert, history, list, multi"
                    })
        except httpx.HTTPStatusError as e:
            logger.error("currency_http_error", status=e.response.status_code)
            return json.dumps({"error": f"API error: HTTP {e.response.status_code}"})
        except Exception as e:
            logger.error("currency_error", error=str(e))
            return json.dumps({"error": str(e)})

    def _run(
        self,
        action: str,
        base: str = "USD",
        target: str = "",
        targets: str = "",
        amount: float = 1.0,
        start_date: str = "",
        end_date: str = "",
    ) -> str:
        coro = self._arun(action, base, target, targets, amount, start_date, end_date)
        try:
            asyncio.get_running_loop()
        except RuntimeError:
            return asyncio.run(coro)

        # If a loop is already running in this thread, run the coroutine in a
        # dedicated worker thread so the sync tool interface still completes.
        with ThreadPoolExecutor(max_workers=1) as executor:
            return executor.submit(asyncio.run, coro).result()
